File: sdk/node/examples/test_loop_commands.ts
/**
 * Test /loop and related slash commands via Node.js SDK.
 * Uses the kimi model from ~/.a3s/config.hcl.
 */

import { Agent } from "@a3s-lab/code";
import { existsSync } from "fs";
import { homedir } from "os";
import { join } from "path";
import { fileURLToPath } from "url";
import { dirname } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

function findConfig(): string {
  const candidates = [
    join(homedir(), ".a3s", "config.hcl"),
    join(__dirname, "..", "..", "..", "..", "..", ".a3s", "config.hcl"),
  ];
  for (const p of candidates) {
    if (existsSync(p)) return p;
  }
  throw new Error(`config.hcl not found in: ${candidates.join(", ")}`);
}

const CONFIG_PATH = findConfig();

function sep(title: string) {
  console.log("=".repeat(60));
  console.log(title);
  console.log("=".repeat(60));
}

async function main() {
  console.log(`Config: ${CONFIG_PATH}\n`);

  const agent = await Agent.create(CONFIG_PATH);
  const session = agent.session("/tmp/test-loop-node");

  // ── 1. listCommands() ──────────────────────────────────────────
  sep("1. listCommands() — all registered slash commands");
  const commands = session.listCommands();
  for (const cmd of commands) {
    const usage = cmd.usage ? `  usage: ${cmd.usage}` : "";
    console.log(`  /${cmd.name.padEnd(15)} ${cmd.description}${usage}`);
  }
  console.log();

  // ── 2. /help via send() ────────────────────────────────────────
  sep("2. /help via send()");
  {
    const r = await session.send("/help");
    console.log(r.text);
  }
  console.log();

  // ── 3. /loop via send() ────────────────────────────────────────
  sep("3. /loop via send() — schedule a 10-second recurring prompt");
  {
    const r = await session.send("/loop 10s say 'tick' and report the time");
    console.log(r.text);
  }
  console.log();

  // ── 4. scheduleTask() — programmatic API ──────────────────────
  sep("4. scheduleTask() — programmatic API (15s interval)");
  const taskId = session.scheduleTask("print current working directory", 15);
  console.log(`Scheduled task ID: ${taskId}`);
  console.log();

  // ── 5. listScheduledTasks() + /cron-list ──────────────────────
  sep("5. listScheduledTasks() + /cron-list");
  const tasks = session.listScheduledTasks();
  console.log(`Active tasks: ${tasks.length}`);
  for (const t of tasks) {
    console.log(
      `  [${t.id}] every ${t.intervalSecs}s — fires in ${t.nextFireInSecs}s — "${t.prompt.slice(0, 50)}"`
    );
  }
  console.log();
  {
    const r = await session.send("/cron-list");
    console.log("/cron-list output:");
    console.log(r.text);
  }
  console.log();

  // ── 6. Real LLM turn (exercises cron drain path) ──────────────
  sep("6. Real LLM turn — exercises cron drain path");
  {
    const r = await session.send("Say hello in one sentence.");
    console.log(`LLM: ${r.text.slice(0, 200)}`);
  }
  console.log();

  // ── 7. /model + /cost + /history (other slash commands) ───────
  sep("7. Other slash commands: /model, /cost, /history");
  for (const cmd of ["/model", "/cost", "/history"]) {
    const r = await session.send(cmd);
    console.log(`${cmd}: ${r.text.split("\n")[0]}`);
  }
  console.log();

  // ── 8. cancelScheduledTask() + /cron-cancel ───────────────────
  sep("8. cancelScheduledTask() + /cron-cancel");
  const cancelled = session.cancelScheduledTask(taskId);
  console.log(`cancelScheduledTask(${taskId}): ${cancelled}`);

  const remaining = session.listScheduledTasks();
  if (remaining.length > 0) {
    const loopId = remaining[0].id;
    const r = await session.send(`/cron-cancel ${loopId}`);
    console.log(`/cron-cancel ${loopId}: ${r.text}`);
  }
  console.log();

  // ── 9. Verify empty ───────────────────────────────────────────
  sep("9. Verify all tasks cancelled");
  const tasksAfter = session.listScheduledTasks();
  console.log(`Remaining tasks: ${tasksAfter.length}`);
  {
    const r = await session.send("/cron-list");
    console.log(r.text);
  }
  console.log();

  sep("All tests passed!");
}

main().catch((e) => {
  console.error("Test failed:", e);
  process.exit(1);
});

File: sdk/python/examples/test_loop_commands.py
#!/usr/bin/env python3
"""Test /loop and related slash commands via Python SDK.

Uses the kimi model from ~/.a3s/config.hcl.
"""

from pathlib import Path


def find_config() -> str:
    # ~/.a3s/config.hcl first, then project .a3s/
    candidates = [
        Path.home() / ".a3s" / "config.hcl",
        Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "config.hcl",
    ]
    for p in candidates:
        if p.exists():
            return str(p)
    raise FileNotFoundError(f"config.hcl not found in: {[str(c) for c in candidates]}")


CONFIG_PATH = find_config()


def find_installed_module():
    try:
        import a3s_code
        return a3s_code
    except ImportError:
        return None


def main():
    mod = find_installed_module()
    if mod is None:
        print("a3s_code not installed — building via maturin develop...")
        import subprocess
        subprocess.run(
            ["maturin", "develop", "--manifest-path",
             str(Path(__file__).parent.parent / "Cargo.toml")],
            check=True
        )
        import a3s_code as mod

    Agent = mod.Agent
    print(f"Config: {CONFIG_PATH}\n")

    agent = Agent.create(CONFIG_PATH)
    session = agent.session("/tmp/test-loop")

    print("=" * 60)
    print("1. list_commands() — all registered slash commands")
    print("=" * 60)
    commands = session.list_commands()
    for cmd in commands:
        usage = f"  usage: {cmd['usage']}" if cmd.get("usage") else ""
        print(f"  /{cmd['name']:15s} {cmd['description']}{usage}")
    print()

    print("=" * 60)
    print("2. /help via send()")
    print("=" * 60)
    result = session.send("/help")
    print(result.text)
    print()

    print("=" * 60)
    print("3. /loop via send() — schedule a 10-second recurring prompt")
    print("=" * 60)
    result = session.send("/loop 10s say 'tick' and report the time")
    print(result.text)
    print()

    print("=" * 60)
    print("4. schedule_task() — programmatic API (15s interval)")
    print("=" * 60)
    task_id = session.schedule_task("print current working directory", 15)
    print(f"Scheduled task ID: {task_id}")
    print()

    print("=" * 60)
    print("5. list_scheduled_tasks() / /cron-list")
    print("=" * 60)
    tasks = session.list_scheduled_tasks()
    print(f"Active tasks: {len(tasks)}")
    for t in tasks:
        print(f"  [{t['id']}] every {t['interval_secs']}s — fires in {t['next_fire_in_secs']}s — \"{t['prompt'][:50]}\"")
    print()
    # Also via send()
    result = session.send("/cron-list")
    print("/cron-list output:")
    print(result.text)
    print()

    print("=" * 60)
    print("6. register_command() — custom Python slash command")
    print("=" * 60)

    def status_handler(args, ctx):
        n_tasks = len(session.list_scheduled_tasks())
        return (
            f"Status: session={ctx['session_id'][:8]}, "
            f"model={ctx['model']}, "
            f"history={ctx['history_len']} msgs, "
            f"scheduled_tasks={n_tasks}, "
            f"args='{args}'"
        )

    session.register_command("status", "Show session status", status_handler)
    result = session.send("/status hello world")
    print(f"/status output: {result.text}")
    print()

    print("=" * 60)
    print("7. Real LLM turn — exercises cron drain path")
    print("=" * 60)
    result = session.send("Say hello in one sentence.")
    print(f"LLM: {result.text[:200]}")
    print()

    print("=" * 60)
    print("8. cancel_scheduled_task() + /cron-cancel")
    print("=" * 60)
    cancelled = session.cancel_scheduled_task(task_id)
    print(f"cancel_scheduled_task({task_id}): {cancelled}")

    remaining = session.list_scheduled_tasks()
    if remaining:
        loop_id = remaining[0]["id"]
        result = session.send(f"/cron-cancel {loop_id}")
        print(f"/cron-cancel {loop_id}: {result.text}")
    print()

    print("=" * 60)
    print("9. Verify all tasks cancelled")
    print("=" * 60)
    tasks_after = session.list_scheduled_tasks()
    print(f"Remaining tasks: {len(tasks_after)}")
    result = session.send("/cron-list")
    print(result.text)
    print()

    print("=" * 60)
    print("All tests passed!")
    print("=" * 60)


if __name__ == "__main__":
    main()
